ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimization.java

```diff
@@ -213,10 +215,17 @@ private boolean secondPreventiveImprovesResults(RaoResult secondPreventiveRaoRes
      */
     private CompletableFuture<RaoResult> postCheckResults(RaoResult raoResult, PrePerimeterResult initialResult, RaoParameters raoParameters) {
         RaoResult finalRaoResult = raoResult;
-        if (raoParameters.getForbidCostIncrease() && raoResult.getCost(OptimizationState.AFTER_CRA) > initialResult.getCost()) {
+
+        double finalCost = finalRaoResult.getCost(OptimizationState.AFTER_CRA);
+        double initialFunctionalCost = initialResult.getFunctionalCost();
+        double initialVirtualCost = initialResult.getVirtualCost();
+        double finalFunctionalCost = finalRaoResult.getFunctionalCost(OptimizationState.AFTER_CRA);
+        double finalVirtualCost = finalRaoResult.getVirtualCost(OptimizationState.AFTER_CRA);
+
+        if (raoParameters.getForbidCostIncrease() && finalCost > initialFunctionalCost + initialVirtualCost) {
             BUSINESS_LOGS.info("RAO has increased the overall cost from {} (functional: {}, virtual: {}) to {} (functional: {}, virtual: {}). Falling back to initial solution:",
-                    formatDouble(initialResult.getCost()), formatDouble(initialResult.getFunctionalCost()), formatDouble(initialResult.getVirtualCost()),
-                    formatDouble(raoResult.getCost(OptimizationState.AFTER_CRA)), formatDouble(raoResult.getFunctionalCost(OptimizationState.AFTER_CRA)), formatDouble(raoResult.getVirtualCost(OptimizationState.AFTER_CRA)));
+                    formatDouble(initialFunctionalCost + initialVirtualCost), formatDouble(initialFunctionalCost), formatDouble(initialVirtualCost),
```

Comment:
i don't understand why you split initialCost in functional + virtual : why not do the same then with finalCost ? In any case, I think it's better when all costs are computed via the same function, i.e getCost inherited function. It seems that getCost can be a bit more elaborate than functional + virtual in case on NaNs.